lib/TPP/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,194 @@
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb = (vnni::utils::isInVnniLayout(memrefB.cast<MemRefType>()))
+                    ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+                    : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+ArrayAttr getUnaryFlags(RewriterBase &rewriter) {
+  return rewriter.getArrayAttr(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+}
+
+ArrayAttr getBinaryFlags(RewriterBase &rewriter, xsmm::BinaryOp addOp) {
+  auto binaryInputType = addOp.getOperand(2).getType().cast<MemRefType>();
+  auto outputType = addOp.getOperand(3).getType().cast<MemRefType>();
+  return rewriter.getArrayAttr(xsmm::BinaryFlagsAttr::get(
+      rewriter.getContext(),
+      xsmm::utils::getBinaryBCast(binaryInputType, outputType,
+                                  /*operandNumber=*/0)));
+}
+
+xsmm::BinaryKindAttr getBinaryKind(RewriterBase &rewriter) {
+  return xsmm::BinaryKindAttr::get(rewriter.getContext(),
+                                   xsmm::BinaryKind::ADD);
+}
+
+xsmm::UnaryKindAttr getUnaryKind(RewriterBase &rewriter) {
+  return xsmm::UnaryKindAttr::get(rewriter.getContext(), xsmm::UnaryKind::RELU);
+}
+
+static ArrayAttr getGemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto brgemmResultOperand = brgemmOp.getOperand(3);
+
+    for (auto use = brgemmResultOperand.user_begin();
+         use != brgemmResultOperand.user_end(); use++) {
+      auto user = *use;
+      if (isa<xsmm::BinaryOp>(user) &&
+          (dyn_cast<xsmm::BinaryOp>(user)).getCallee() ==
+              mlir::xsmm::BinaryKind::ADD) {
```

Comment:
It might be safer to break fusion on unknown users.
Or maybe it is handled well, just make sure to add some test cases to cover e.g., a copy in between XSMM calls.

Reply:
will address this in a test case in the next commit

Reply:
If there's anything untested, it must return failure and not transform.